File: SparkRDD/Java/Tweets/LongestTweet.java
package org.masterbigdata3;

import org.apache.log4j.Level;
import org.apache.log4j.Logger;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import scala.Tuple2;

import java.util.List;

public class LongestTweet {


    public static void main(String[] args) {
        Logger.getLogger("org").setLevel(Level.OFF);

        SparkConf sparkConf = new SparkConf() // We create the SparkConf
                .setAppName("LogestTweet_Spark")
                .setMaster("local[2]");

        JavaSparkContext sparkContext = new JavaSparkContext(sparkConf); // and the SparkContext

        // Read the file
        JavaRDD<String> lines = sparkContext.textFile("/home/master/Descargas/tweets.txt");

        // Separate the columns
        JavaRDD<String[]> fields = lines.map(line -> line.split("\t"));

        // Take the columns we want
        JavaPairRDD<String, Integer> longestTweet = fields.mapToPair(array -> new Tuple2<>(array[1] + " - " + array[3], array[2].length())); //          .map(lambda array: (array[1], len(array[2]), array[3])) \

        // Sort the RDD
        JavaPairRDD<Integer, String> longesTweetSorted = longestTweet.mapToPair(pair -> new Tuple2<>(pair._2, pair._1)).sortByKey(false);

        // Take the 20 longest tweets
        List<Tuple2<Integer, String>> output = longesTweetSorted.take(20);

        // Now we can show it
        for (Tuple2<?, ?> tuple : output) {
            System.out.println(tuple._2() + ", " + tuple._1());
        }
    }
}

File: SparkRDD/Python/Tweets/LongestTweet.py
"""find the longest tweet emitted by any user.
The program will print in the screen the user name,
the length of the tweet, and the time and date of the tweet."""

import sys
from pyspark import SparkConf, SparkContext

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: spark-submit LongestTweet.py <file>", file=sys.stderr)
        exit(-1)

    # Create the SparkConf and the SparkContext
    spark_conf = SparkConf().setAppName("LongestTweet").setMaster("local[2]")
    spark_context = SparkContext(conf=spark_conf)

    # Create the logger
    logger = spark_context._jvm.org.apache.log4j
    logger.LogManager.getLogger("org").setLevel(logger.Level.WARN)

    # \t is for tabulations
    tweetListSorted = spark_context \
        .textFile(sys.argv[1]) \
        .map(lambda line: line.split('\t')) \
        .map(lambda array: (array[1], len(array[2]), array[3])) \
        .sortBy(lambda array: array[1], False)

    # Take the tweet which has more characters
    tweetMax = tweetListSorted.first()

    # Take all the tweets which have the same characters than the previous tweet
    output = tweetListSorted.filter(lambda tweet: tweet[1] == tweetMax[1]).collect()

    # Show it
    print("Name,      ", "Length of tweet,              ", "Date")
    for word in range(output.__len__()):
        print(output[word][0], ": ", output[word][1], " - " ,output[word][2])

    # Stop the Sparkcontext
    spark_context.stop()